### DISCUSSION TITLE
Large number of 301 redirects

### DISCUSSION THREAD

**Discussion by PrimozRome** [vercel/next.js#40219] [Help]
Hi there. I need to handle large sum of redirects for our new corporate website which will run on the Next.js and the Vercel CDN. I need to handle about 4000 redirects trough multiple language locales. Is using the redirects key in `next.config.js` good way to do this large sum of redirects? I won't be able to use Regex expressions so I will have to define each redirect separately. Is there something else that I can setup these 301 redirects?

**Comment by icyJoseph** [comment]
Middleware is another option, you'd even be able to test the redirects.

I'd be a bit worried about the size of the file though, I think there's a limit to how large the middleware bundle can be, and about 4000 redirects might cross that limit. 

Otherwise, I'd say the `next.config.js` file should be alright, some people define this properties in other modules and then require them into the `next.config.js` module, so that things are kept clean and tidy.

I am not sure if having another layer handle this is ever possible for your use case, or architecture, but it feels like redirecting this much should be done by NGINX, or any of those server running engines, instead of the frontend rendering engine, but that's just a thought.

**Comment by PrimozRome** [comment]
>I am not sure if having another layer handle this is ever possible for your use case, or architecture, but it feels like redirecting this much should be done by NGINX, or any of those server running engines, instead of the frontend rendering engine, but that's just a thought.

Yeah I definitely agree, but how to do that on Vercel...?

**Reply by icyJoseph** [reply]
I guess, it is possible, but reading through this post https://github.com/vercel/vercel/discussions/5769, it looks like it is discouraged, I'd go with your proposal of `next.config.js`, or middleware for now. See how that scales to 4000 redirects.

**Comment by rangler2** [comment]
@PrimozRome   I have a similar requirement, would be good if you can share how it went? FYI  I have just done this with just over 1000 page-to-page redirects in next.config.js  seems to work OK.
